tests/pytorch/test_fused_attn.py

```diff
@@ -0,0 +1,624 @@
+        amax_history_len=1,
+        amax_compute_algo="most_recent",
+    )
+
+    dpa = DPA_FP8(config).to(dtype = torch.float16).cuda()
+    with fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):
+        op = dpa(inp, cu_seqlens, config.seq_len)
+        op.backward(op_grad)
+
+    context = torch.load("ctx.pt")
+    dqkv = torch.load('dqkv.pt')
+    return (context.view(bs, config.seq_len, -1).transpose(0,1),
+        dqkv.view(bs, config.seq_len, 3, config.num_attention_heads, config.head_dim).transpose(0,1).contiguous())
+
+def _run_dpa_fp8_ref(dtype, bs, config, backend):
+
+    os.environ["NVTE_FLASH_ATTN"] = "0"
+    os.environ["NVTE_FUSED_ATTN"] = "0"
+    if backend == "FlashAttention":
+        os.environ["NVTE_FLASH_ATTN"] = "1"
+    if backend == "FusedAttention":
+        os.environ["NVTE_FUSED_ATTN"] = "1"
+
+    inp = torch.load('qkv.pt').cuda()
+    inp.requires_grad=True
+    seqlens = torch.empty(bs, dtype = torch.int32).cuda()
+    seqlens.fill_(config.seq_len)
+    cu_seqlens = torch.zeros(bs + 1, device = inp.device, dtype = torch.int32)
+    cu_seqlens[1:] = torch.cumsum(seqlens, dim = 0)
+    op_grad = torch.load('op_grad.pt').cuda().view(bs, config.seq_len, -1).transpose(0,1)
+
+    block = (
+         DotProductAttention(
+                config.num_attention_heads,
+                config.head_dim,
+                attention_dropout = config.dropout_p,
+                attn_mask_type = config.attn_mask_type,
+                sequence_parallel = False,
+                tp_size = 1,
+                get_rng_state_tracker = None,
+                tp_group = None,
+                layer_number = 1,
+                attention_type = "self"
+        ).to(dtype = dtype).cuda()
+    )
+
+    q = inp[:, :,0,:,:]
+    k = inp[:, :,1,:,:]
+    v = inp[:, :,2,:,:]
+    op = block(q, k, v)
+    op.backward(op_grad)
+    torch.save(op,'ctx_ref.pt')
+    torch.save(inp.grad,'dqkv_ref.pt')
+
+    return op, inp.grad
+
+from torch.nn.parameter import Parameter
+import transformer_engine.pytorch.cpp_extensions as ext
+import transformer_engine_extensions as tex
+import transformer_engine.pytorch.fp8 as fp8
+from transformer_engine.pytorch import fp8_autocast
+from transformer_engine.pytorch.module.base import TransformerEngineBaseModule, _prepare_backward
+from transformer_engine.common import recipe
+from typing import Union, Dict, Any, Tuple, List
+from transformer_engine.pytorch.cpp_extensions import fused_attn_fwd_qkvpacked, fused_attn_bwd_qkvpacked
+from transformer_engine.pytorch.cpp_extensions import FusedAttnBackend
+
+_CUBLASLT_WORKSPACE_SIZE_BYTES = 33_554_432  # 32MiB
+_2X_ACC_FPROP = False
+_2X_ACC_DGRAD = False
+_2X_ACC_WGRAD = False
+
+META_QKV  = tex.FP8FwdTensors.GEMM1_OUTPUT
+META_O    = tex.FP8FwdTensors.GEMM2_INPUT
+META_DO   = tex.FP8BwdTensors.GRAD_INPUT2
+META_DQKV = tex.FP8BwdTensors.GRAD_OUTPUT1
+
+META_S    = 7 #tex.FP8FwdTensors.GEMM2_OUTPUT
+META_DS   = 5 #tex.FP8BwdTensors.GRAD_INPUT1
```

Comment:
Why hardcode these?

Reply:
These two enum keys are not related to gemms. As long as they're not conflicting with the two 'real' gemms in fp8 mha (i.e. fwd 0-5 and bwd 0-3), they're fine. Would you recommend adding a set of GEMM3 keys to tex.FP8FwdTensors and tex.FP8BwdTensors?

Reply:
I added GEMM3 to these enums for now. Maybe in the future we have a better way to expand these enums instead of altering the code.